src/h3/mod.rs

```diff
@@ -26,4 +26,940 @@
+    peer_control_stream_open: bool,
+
+    qpack_encoder: qpack::Encoder,
+    qpack_decoder: qpack::Decoder,
+
+    qpack_encoder_stream_open: bool,
+    peer_qpack_encoder_stream_open: bool,
+    qpack_decoder_stream_open: bool,
+    peer_qpack_decoder_stream_open: bool,
+}
+
+impl Connection {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(config: &mut Config, is_server: bool) -> Result<Connection> {
+        let root = String::clone(&config.root_dir); // TODO shouldn't need to clone here
+
+        Ok(Connection {
+            root_dir: root,
+            is_server,
+            requests: RequestMap::new(),
+            streams: StreamMap::new(),
+            uni_stream_ledger: Vec::new(),
+
+            num_placeholders: config.num_placeholders,
+            max_header_list_size: config.max_header_list_size,
+            qpack_max_table_capacity: config.qpack_max_table_capacity,
+            qpack_blocked_streams: config.qpack_blocked_streams,
+
+            peer_num_placeholders: None,
+            peer_max_header_list_size: None,
+            peer_qpack_max_table_capacity: None,
+            peer_qpack_blocked_streams: None,
+
+            control_stream_id: None,
+            peer_control_stream_open: false,
+
+            qpack_encoder: qpack::Encoder::new(),
+            qpack_decoder: qpack::Decoder::new(),
+
+            qpack_encoder_stream_open: false,
+            peer_qpack_encoder_stream_open: false,
+            qpack_decoder_stream_open: false,
+            peer_qpack_decoder_stream_open: false,
+        })
+    }
+
+    /// Get a request stream ID if there is one available
+    pub fn get_available_request_stream(&mut self, req: http::Request<()>) -> Result<u64> {
+        if self.requests.is_empty() {
+            self.requests.insert(0, req);
+            Ok(0)
+        } else {
+            let id = self.streams.keys().last().unwrap();
+
+            // TODO: this should check the value of the connection MAX_STREAM_ID
+            if id < &std::u64::MAX {
+                let id = id + 4;
+                Ok(id)
+            } else {
+                Err(Error::LimitExceeded)
+            }
+        }
+    }
+
+    /// Returns an available stream ID for the local endpoint to use
+    fn get_available_uni_stream(&mut self) -> Result<u64> {
+        if self.uni_stream_ledger.is_empty() {
+            if self.is_server {
+                self.uni_stream_ledger.push(0x3);
+            } else {
+                self.uni_stream_ledger.push(0x2);
+            }
+
+            Ok(*self.uni_stream_ledger.last().unwrap())
+        } else {
+            let id = self.uni_stream_ledger.last().unwrap();
+
+            // TODO: this should check the value of the connection MAX_STREAM_ID
+            if id < &std::u64::MAX {
+                let id = id + 4;
+                Ok(id)
+            } else {
+                Err(Error::LimitExceeded)
+            }
+        }
+    }
+
+    pub fn is_established(&self) -> bool {
```

Comment:
This name seems confusing. "established" to me means that the handshake is complete, but this only checks if the local streams have been opened, can it be renamed to something else? e.g. `is_open()`.

Reply:
The function now also checks if the peer opened streams and you received/processed them e.g. Control stream SETTINGS. This is not perfect so leaving open for further discussion.